cpp/src/arrow/type.h

```diff
@@ -1546,6 +1581,97 @@ std::shared_ptr<Schema> schema(
 
 /// @}
 
+/// \brief Convenience class to incrementally construct/merge schemas.
+///
+/// This class amortizes the cost of validating field name conflicts by
+/// maintaining the mapping. The caller also controls the conflict resolution
+/// scheme.
+class SchemaBuilder {
+ public:
+  // Indicate how field conflict(s) should be resolved when building a schema. A
+  // conflict arise when a field is added to the builder and one or more field(s)
+  // with the same name already exists.
+  enum ConflictPolicy {
+    // Ignore the conflict and append the field. This is the default behavior of the
+    // Schema constructor and the `arrow::schema` factory function.
+    APPEND = 0,
+    // Replace the existing field with the newer one.
+    REPLACE,
+    // Merge the fields. See documentation of `Field::WithField`.
+    MERGE,
+    // Refuse the new field and error out.
+    ERROR,
+  };
+
+  /// \brief Construct an empty SchemaBuilder
+  explicit SchemaBuilder(ConflictPolicy conflict_policy = APPEND);
+  /// \brief Construct an SchemaBuilder from a list of fields
+  SchemaBuilder(std::vector<std::shared_ptr<Field>> fields,
+                ConflictPolicy conflict_policy = APPEND);
+  /// \brief Construct an SchemaBuilder from a schema, preserving the metadata
+  SchemaBuilder(const std::shared_ptr<Schema>& schema,
+                ConflictPolicy conflict_policy = APPEND);
+
+  /// \brief Return the conflict resolution method.
+  ConflictPolicy policy() const { return policy_; }
+
+  /// \brief Set the conflict resolution method.
+  void SetPolicy(ConflictPolicy resolution) { policy_ = resolution; }
+
+  /// \brief Add a field to the constructed schema.
+  ///
+  /// \param[in] field to add to the constructed Schema.
+  /// \return A failure if encountered.
+  Status WithField(const std::shared_ptr<Field>& field);
+
+  /// \brief Add multiple fields to the constructed schema.
+  ///
+  /// \param[in] fields to add to the constructed Schema.
+  /// \return The first failure encountered, if any.
+  Status WithFields(const std::vector<std::shared_ptr<Field>>& fields);
+
+  /// \brief Add fields of a Schema to the constructed Schema.
+  ///
+  /// \param[in] schema to take fields to add to the constructed Schema.
+  /// \return The first failure encountered, if any.
+  Status WithSchema(const std::shared_ptr<Schema>& schema);
+
+  /// \brief Add fields of multiple Schemas to the constructed Schema.
+  ///
+  /// \param[in] schemas to take fields to add to the constructed Schema.
+  /// \return The first failure encountered, if any.
+  Status WithSchemas(const std::vector<std::shared_ptr<Schema>>& schemas);
+
+  Status WithMetadata(const KeyValueMetadata& metadata);
```

Comment:
`AddMetadata`. Other `WithMetadata` methods return a _new_ instance with added metadata, leaving `this` intact.